File: frameworks/java/javalin/UnitTests/src/test/java/unittests/StreamTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import io.javalin.testtools.JavalinTest;
import io.javalin.testtools.Response;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

class StreamTests extends JavalinApp {

    // rb:test stream.ndjson
    @Test
    @Tag("stream.ndjson")
    void eachRowIsALineAndTheLengthIsNeverSent() {
        JavalinTest.test(app(), (server, client) -> {
            Response response = client.get("/stream/items");

            assertTrue(Answer.header(response, "content-type").startsWith("application/x-ndjson"));
            assertNull(Answer.header(response, "content-length"));
            String[] lines = response.body().string().stripTrailing().split("\n");
            JsonNode rows = Expected.json("items.medium.json").get("items");
            assertEquals(rows.size(), lines.length);
            for (int i = 0; i < lines.length; i++) {
                assertEquals(rows.get(i), Expected.JSON.readTree(lines[i]));
            }
        });
    }
}

File: frameworks/dotnet/hardened/UnitTests/StreamTests.cs
using Hardened.Web.Kestrel.Runtime;

namespace UnitTests;

/// <summary>On a Kestrel socket, so the headers are what the server framed the stream with.</summary>
[KestrelRuntime]
public class StreamTests
{
    // rb:test stream.ndjson
    [ModuleTest]
    [Trait("corpus", "stream.ndjson")]
    public async Task Each_row_is_a_line_and_the_length_is_never_sent(ITestWebApp app)
    {
        TestWebResponse response = await app.Get("/stream/items");

        response.Assert.Ok();
        Assert.StartsWith("application/x-ndjson", Answer.Header(response, "Content-Type"));
        Assert.Equal("chunked", Answer.Header(response, "Transfer-Encoding"));
        string[] lines = (await response.ReadTextAsync()).TrimEnd('\n').Split('\n');
        JsonArray rows = Expected.Json("items.medium.json")["items"]!.AsArray();
        Assert.Equal(rows.Count, lines.Length);
        Assert.All(lines.Zip(rows), pair => Assert.True(JsonNode.DeepEquals(pair.Second, JsonNode.Parse(pair.First))));
    }
}
